### PR TITLE
Foreign key constraints

### PR DESCRIPTION
**Pull request by eudaimonious**
We didn't have any foreign key constraints in our database. I read through schema.rb to uncover columns that were foreign keys. I used the add_foreign_key method to add them to postgres. This method isn't available until Rails 4.2 so I added the Foreigner gem to support it.

https://github.com/studentinsights/studentinsights/issues/209

### CODE REVIEW

**Comment by eudaimonious** [comment]
closing and re-opening to re-trigger travis

**Comment by alexsoble** [comment]
Thanks for this @eudaimonious!

Helped me learn something new, your PR made me go and read up on foreign keys: https://robots.thoughtbot.com/referential-integrity-with-foreign-keys

My only suggestion here would be, why don't we bump up the Rails version? We're very close to 4.2, and that way we have one less dependency, which is nice. 

Anyway, that's all I've got. Excellent work!

**Comment by kevinrobinson** [comment]
@eudaimonious This is awesomely helpful, thank you! :+1:

File: db/migrate/20160330000130_add_foreign_key_constraints.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
+class AddForeignKeyConstraints < ActiveRecord::Migration
+  def change
+    add_foreign_key :absences, :student_school_years
```
Comment (kevinrobinson) [inline db/migrate/20160330000130_add_foreign_key_constraints.rb]: From looking at https://robots.thoughtbot.com/referential-integrity-with-foreign-keys, it seems we can also move `dependent: destroy` expressions here too.  I don't think we're super rigorous about that at the moment, but would be something good to try.

Also, there's a recommendation about verifying first that the current state of the production DB can satisfy these contraints. :)  We can try that out on the staging site, which should be a quick way to validate before deploying this for real.  Added https://github.com/studentinsights/studentinsights/issues/246 for future us.

**Comment by kevinrobinson** [comment]
I tried deploying this to staging and then running the migration.  The migration was successful, but then I ran into errors that I think are completely unrelated but didn't investigate further yet.

```
2016-03-31T02:57:19.110486+00:00 app[web.1]: Started GET "/homerooms/7" for 98.216.106.247 at 2016-03-31 02:57:19 +0000
2016-03-31T02:57:19.112216+00:00 app[web.1]: Processing by HomeroomsController#show as HTML
2016-03-31T02:57:19.112234+00:00 app[web.1]:   Parameters: {"id"=>"7"}
2016-03-31T02:57:19.137306+00:00 app[web.1]: Completed 500 Internal Server Error in 25ms (ActiveRecord: 10.5ms)
2016-03-31T02:57:19.138299+00:00 app[web.1]:
2016-03-31T02:57:19.138307+00:00 app[web.1]: NoMethodError (undefined method `id' for nil:NilClass):
2016-03-31T02:57:19.138308+00:00 app[web.1]:   app/models/intervention.rb:53:in `most_recent_atp'
2016-03-31T02:57:19.138309+00:00 app[web.1]:   app/models/student.rb:204:in `most_recent_atp_hours'
2016-03-31T02:57:19.138310+00:00 app[web.1]:   app/controllers/homerooms_controller.rb:49:in `fat_student_hash'
2016-03-31T02:57:19.138310+00:00 app[web.1]:   app/controllers/homerooms_controller.rb:10:in `block in show'
2016-03-31T02:57:19.138311+00:00 app[web.1]:   app/controllers/homerooms_controller.rb:10:in `show'
```

I suspect this is from the staging site not being in a great state somehow, and not actually related to this migration, but since it takes a few hours to rebuild the staging site I can't try now.  I think the better path is to just ship this and allocate time after to validate and fix forward if anything does come up.

File: db/migrate/20160330000130_add_foreign_key_constraints.rb (hunk 1 above)
Comment (eudaimonious) [inline db/migrate/20160330000130_add_foreign_key_constraints.rb]: Agreed. This wouldn't be an across the board change you'd want to make for all instances of a foreign key so as someone who doesn't know all of the business logic yet I guess I'd just base the addition on whether I see a belongs_to relationship or a presence validation.....

**Comment by eudaimonious** [comment]
@alexsoble Upgrading to 4.2 and removing the gem could be a good code quality next step. Or hey, maybe we want to go somewhere past 4.2.....

@kevinrobinson I'm not sure if it's related (if it is we should fix before merging) but the InterventionType class that the failing method refers to seems ultra suspect. The first line in the file is `#Deprecated`

**Comment by eudaimonious** [comment]
Thank you for such a thoughtful review!

File: db/migrate/20160330000130_add_foreign_key_constraints.rb (hunk 1 above)
Comment (kevinrobinson) [inline db/migrate/20160330000130_add_foreign_key_constraints.rb]: :+1: yeah I think it has to be a manual walk through the semantics of each relationship, I don't think anyone knows all the relationships off the top of their head (@alexsoble might still!).  Not sure I have any better recommendation but am happy to be another set of eyes, so if that's helpful ever feel free to ping me!

**Comment by kevinrobinson** [comment]
@eudaimonious Yep, that sounds like a great plan!

FWIW, `InterventionType` is only deprecated in the last month (in that the UI for the write path has been removed).  It's still exercised on some older code on the read path (that can be removed in another week or two for https://github.com/studentinsights/studentinsights/issues/211), so I don't think there's any other reason to suspect that it's a code bug there causing issues.  There is also known drift at the moment between staging and production datasets (https://github.com/studentinsights/studentinsights/issues/9), so I think the thing to do is rebuild the staging site and then see if we can still reproduce.  I'm guessing things will work smoothly there, or else digging into it will reveal some other drift, but only one way to find out. :)

Sorry to hold this up, but since it requires validating production data, I'll try to kick it off tonight and then see where we're at tomorrow or in the next few days.

**Comment by kevinrobinson** [comment]
I rebuilt the staging site, deployed this and ran the migration, and things look good! :+1: :ship:

**Comment by kevinrobinson** [comment]
This failed on production on `StudentSchoolYear` pointing to a Student record that didn't exist.  Going to revert this, https://github.com/studentinsights/studentinsights/issues/264 has the next steps for investigation.

**Pull request merged**